### ISSUE TITLE
iOS/Metal: Depth buffer fighting

### ISSUE THREAD

**Issue by BStringhamVRSK** [google/filament#6478]
**On iOS/Metal we are getting depth-buffer conflicts**
We are porting our app to iOS from Windows.  They share the same code which controls the Filament library calls, so the results should be nearly identical.
We are getting depth-buffer conflicts when we draw the same polygon twice--once with an opaque material, and again with an alpha-transparent material.  We do not see this problem on Windows/OpenGL

See the screenshots for comparison between Windows/OpenGL and iOS/Metal.

Since the DepthFunc defaults to RasterState::DepthFunc::GE (which is desired), I don't expect depth conflicts, since the depths of both writes should be identical (same polygon/vertex buffer for both draws).

Is the default render state slightly different between OpenGL and Metal, so I need to make some explicit call to correct this?

**To Reproduce**
1. Draw a polygon with an opaque material
2. Draw it again with a translucent material
3. Observe the depth-buffer conflicts

**Screenshots**
**iOS/Metal - depth buffer conflict**
![iOS_Metal_Cropped](https://user-images.githubusercontent.com/40642629/213582639-2d10a6da-33aa-4170-9dce-7978070e9bff.png)

**Windows/OpenGL**
![Windows_OpenGL](https://user-images.githubusercontent.com/40642629/213582662-117caa78-8456-43a3-b6c8-fdeeed0d473a.PNG)

**Smartphone (please complete the following information):**
 - Device:  iPhone 13 Pro Max
 - OS:  iOS 16.2

**Comment by romainguy** [comment]
> I don't expect depth conflicts, since the depths of both writes should be identical (same polygon/vertex buffer for both draws)

Unfortunately the depth is not guaranteed to be exactly equal across draw calls, in large part because the shader compiler might re-order some of the computations which, because of floating point imprecisions, will lead to slightly different results. The only way to handle this correctly across APIs/GPUs is to use a polygon offset for your translucent draw call.

Note that you would also see this with GL/Windows, depending on the GPU/drivers you are using. We ran into this a lot when trying to support a depth prepass, and we gave up in part because it was extremely difficult to get consistent depth results across OS/GPU/APIs/drivers.

**Comment by BStringhamVRSK** [comment]
> Unfortunately the depth is not guaranteed to be exactly equal across draw calls, in large part because the shader compiler might re-order some of the computations which, because of floating point imprecisions, will lead to slightly different results.
> 

I don't mean to second-guess your answer because you know your system much better than I do, but here is a suggestion:  I had to solve this problem a few years ago while creating a deferred renderer which required exact position/depth results across multiple passes.  Maybe you've already pursued these options:

GLSL and HLSL use the _**precise**_ keyword to address the problem.
MSL uses the _**invariant**_ keyword.
I haven't investigated Vulkan's solution.

Another related issue I encountered on old Xbox hardware was its ability to perform sub-pixel offsets (I don't remember what they called it).  I had to ensure that the offset was consistent across shaders.
